queries/stack-graphs.tsg

```diff
@@ -0,0 +1,818 @@
+  (while_let_expression value: [(identifier) (self)] @ident)
+  (yield_expression [(identifier) (self)] @ident)
+
+  ;; ### Types
+  (array_type length: [(identifier) (self)] @ident)
+
+  ;; ### Statements
+  (block (_)* [(identifier) (self)] @ident)
+  (expression_statement [(identifier) (self)] @ident)
+  (let_declaration value: [(identifier) (self)] @ident)
+
+  ;; ### Items
+  (const_item value: [(identifier) (self)] @ident)
+  (static_item value: [(identifier) (self)] @ident)
+] {
+  node @ident.use
+  attr (@ident.use) node_reference = @ident
+  edge @ident.use -> @ident.scope
+}
+
+;; List all contexts where an identifier is a reference to a binding in the
+;; "types" namespace.
+[
+  ;; ### Types
+  (abstract_type trait: (type_identifier) @ident)
+  (array_type element: (type_identifier) @ident)
+  (bounded_type (type_identifier) @ident)
+  (dynamic_type trait: (type_identifier) @ident)
+  (function_type trait: (type_identifier) @ident)  ;; only for Fn, FnMut, FnOnce
+  (function_type return_type: (type_identifier) @ident)
+  (generic_type type: (type_identifier) @ident)
+  (generic_type_with_turbofish type: (type_identifier) @ident)
+  (pointer_type type: (type_identifier) @ident)
+  (reference_type type: (type_identifier) @ident)
+  ;; Scoped identifiers are more complicated than that, so comment them out for now.
+  ;; (Also the scoped_identifier nodes in use_declaration nodes should not match.)
+  ;; Definitely lexical scopes have a types/values split; what about other namespaces,
+  ;; like the one under a trait, a struct, a module...?
+  ;;(scoped_identifier path: (identifier) @ident)
+  (tuple_type (type_identifier) @ident)
+  (type_arguments (type_identifier) @ident) ;; questionable
+
+  ;; ### Expressions
+  (type_cast_expression type: (type_identifier) @ident)
+  (bracketed_type (type_identifier) @ident)
+  (struct_expression name: (type_identifier) @ident)
+
+  ;; ### Patterns
+  (struct_pattern type: (type_identifier) @ident)
+
+  ;; ### Items
+  (const_item type: (type_identifier) @ident)
+  (constrained_type_parameter left: (type_identifier) @ident)
+  (field_declaration type: (type_identifier) @ident)
+  (function_item return_type: (type_identifier) @ident)
+  (impl_item trait: (type_identifier) @ident)
+  (impl_item type: (type_identifier) @ident)
+  (ordered_field_declaration_list type: (type_identifier) @ident)
+  (parameter type: (type_identifier) @ident)
+  (static_item type: (type_identifier) @ident)
+  (trait_bounds (type_identifier) @ident)
+  (where_predicate left: (type_identifier) @ident)
+] {
+  node @ident.use
+  attr (@ident.use) node_reference = @ident
+
+  let t = (node)
+  attr (t) push_symbol = "%type"
+  edge @ident.use -> t
+  edge t -> @ident.scope
+}
+
+;; The type-tag in a tuple_struct_pattern is a gruesome special case in rustc.
+;; It looks in both namespaces, preferring inner scopes to outer scopes, and
+;; when it finds both a value and a type in the same lexical scope, picking the
+;; type.
```

Comment:
I tried to understand this one, but didn't manage yet. What I do for tricky cases that I cannot/do not support is write a test that shows the behavior and name it `.rs.skip`. This way it doesn't run, but it's still there as documentation.